File: services/customer/internal/domain/service/payment_service.go
package service

import (
	"context"

	"github.com/codejsha/bookstore-microservices/customer/internal/application/port/pb/paymentpb"
	"github.com/codejsha/bookstore-microservices/customer/internal/application/usecase"
	"github.com/codejsha/bookstore-microservices/customer/internal/domain/aggregate"
	"github.com/codejsha/bookstore-microservices/customer/internal/infrastructure/adapter/protostub"
)

var _ usecase.PaymentUseCase = (*paymentService)(nil)

func NewPaymentService(
	paymentGrpcClient *protostub.PaymentGrpcClient,
) usecase.PaymentUseCase {
	return &paymentService{
		paymentServiceClient: paymentGrpcClient.Client,
	}
}

type paymentService struct {
	paymentServiceClient paymentpb.PaymentServiceClient
}

func (s paymentService) FindAllCustomerPayments(ctx context.Context, req *paymentpb.PaymentFindAllProtoReq) (int64, []*aggregate.PaymentAggregate, error) {
	// find payments
	resp, err := s.paymentServiceClient.FindAllPayments(ctx, req)
	if err != nil {
		return 0, nil, err
	}

	// return response
	paymentAggs := make([]*aggregate.PaymentAggregate, 0, len(resp.Items))
	for i, item := range resp.Items {
		paymentAggs[i] = aggregate.NewPaymentAggregate(item)
	}
	return resp.Total, paymentAggs, nil
}

func (s paymentService) FindCustomerPayment(ctx context.Context, req *paymentpb.PaymentFindProtoReq) (*aggregate.PaymentAggregate, error) {
	// find payment
	resp, err := s.paymentServiceClient.FindPayment(ctx, req)
	if err != nil {
		return nil, err
	}
	return aggregate.NewPaymentAggregate(resp), nil
}

File: services/admin/src/main/kotlin/com/codejsha/bookstore/admin/domain/service/PaymentService.kt
package com.codejsha.bookstore.admin.domain.service

import com.codejsha.bookstore.admin.application.port.restclient.PaymentClient
import com.codejsha.bookstore.admin.application.usecase.PaymentUseCase
import com.codejsha.bookstore.admin.domain.model.option.PaymentQueryOption
import com.codejsha.bookstore.admin.domain.model.option.RefundQueryOption
import com.codejsha.platform.shared.data.ActorContext
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service

@Service
class PaymentService(
    private val paymentClient: PaymentClient,
) : PaymentUseCase {

    override suspend fun findAllPayments(option: PaymentQueryOption, pageable: Pageable, context: ActorContext) =
        paymentClient.findAllPayments(option, pageable)

    override suspend fun findPayment(uid: String, context: ActorContext) = paymentClient.findPayment(uid)

    override suspend fun findAllRefunds(option: RefundQueryOption, pageable: Pageable, context: ActorContext) =
        paymentClient.findAllRefunds(option, pageable)

    override suspend fun findRefund(uid: String, context: ActorContext) = paymentClient.findRefund(uid)
}
